dipy/denoise/noise_estimate.py

```diff
@@ -76,6 +80,101 @@ def piesno(data, N=1, alpha=0.01, l=100, itermax=100, eps=1e-5, return_mask=Fals
     Journal of Magnetic Resonance 2009; 197: 108-119.
     """
 
+    # This method works on a 2D array with repetitions as the third dimension,
+    # so process the dataset slice by slice.
+    if data.ndim < 3:
+        e_s = "This function only works on datasets of at least 3 dimensions."
+        raise ValueError(e_s)
+
+    if data.ndim == 4:
+
+        sigma = np.zeros(data.shape[-2], dtype=np.float32)
+        mask_noise = np.zeros(data.shape[:-1], dtype=np.bool)
+
+        for idx in range(data.shape[-2]):
+            sigma[idx], mask_noise[..., idx] = _piesno_3D(data[..., idx, :], N,
+                                                          alpha=alpha,
+                                                          l=l,
+                                                          itermax=itermax,
+                                                          eps=eps,
+                                                          return_mask=True)
+
+    else:
+        sigma, mask_noise = _piesno_3D(data, N,
+                                       alpha=alpha,
+                                       l=l,
+                                       itermax=itermax,
+                                       eps=eps,
+                                       return_mask=True)
+
+    if return_mask:
+        return sigma, mask_noise
+
+    return sigma
+
+
+def _piesno_3D(data, N, alpha=0.01, l=100, itermax=100, eps=1e-5,
+               return_mask=False):
```

Comment:
Yes - I think this is fine. This is a 'private' function anyhow, so users are discouraged from using it directly.